### ISSUE TITLE
Using AddRange on Lst multiple times causes strange results

### ISSUE THREAD

**Issue by colethecoder** [louthy/language-ext#675]
If you have 2 Lsts and you use AddRange to concatenate them it works correctly the first time but subsequent AddRange execution causes unexpected results:

```cs
            var l1 = LanguageExt.Prelude.List<int>(1, 2, 3);
            var l2 = LanguageExt.Prelude.List<int>(4, 5, 6);

            var a = l1.AddRange(l2); // Count 6, [1,2,3,4,5,6]
            var b = l1.AddRange(l2); // Count 5, [1,2,4,5,6]
            var c = l1.AddRange(l2); // Count 8, [1,2,4,5,6,4,5,6]
            var d = l1.AddRange(l2); // Count 7, [1,2,4,5,4,5,6]
            var e = l1.AddRange(l2); // Count 6, [1,2,4,4,5,6]
```

**Comment by colethecoder** [comment]
I assume this is similar to the previous Seq bugs, I will investigate further.

**Comment by louthy** [comment]
> I assume this is similar to the previous Seq bugs, I will investigate further.

I doubt it, it uses a completely different backing storage mechanism.  Let me know what you find anyway 👍

**Comment by louthy** [comment]
@colethecoder It will be to do with the `Pool` in `ListEnumerator`

**Comment by bender2k14** [comment]
The problem is that `Lst<>` is not immutable.  @colethecoder found a case in which `AddRange` mutates one of its inputs.  Here is a simpler tests that demonstrates this.
```C#
[Fact]
public void AddRange_TwoLengthThreeLists_InputsUnchanged()
{
    var actualLeft = List(1, 2, 3);
    var actualRight = List(4, 5, 6);
    var expectedLeft = Seq(1, 2, 3);
    var expectedRight = Seq(4, 5, 6);
    Assert.Equal(expectedLeft, actualLeft);
    Assert.Equal(expectedRight, actualRight);

    _ = actualLeft.AddRange(actualRight);

    Assert.Equal(expectedLeft, actualLeft); // fails here with actualLeft = [1, 2]
    Assert.Equal(expectedRight, actualRight);
}
```

`Lst<>` depends on `LstInternal<>` which depends on `ListItem<>` which has four mutable fields:
https://github.com/louthy/language-ext/blob/e668a9b94a5fb5714785fced2a2983f548f4cc27/LanguageExt.Core/DataTypes/List/Lst.Internal.cs#L458-L466

In F#, the immutable `list` data structure is a linked list.  In Language Ext, `LstInternal<>` is a [self-balancing binary tree](https://en.wikipedia.org/wiki/Self-balancing_binary_search_tree) with each `ListItem<>` instance being a node in a tree.  However, I don't know which one of the many [implementations](https://en.wikipedia.org/wiki/Self-balancing_binary_search_tree#Implementations) is intended.

@louthy, I think you sensed that there was an issue here.  When you were [optimizing this code back in the summer of 2018 to improve runtime efficiency](https://github.com/louthy/language-ext/commit/996743d8d26b77db8d5573623040cb8b437ded8e#diff-e3716380c0d06081191c51f2f160ec27R458-R622), you also committed commented out code that seems like it would have constructed new `ListItem<>` instances instead of mutating the existing ones.
https://github.com/louthy/language-ext/blob/e668a9b94a5fb5714785fced2a2983f548f4cc27/LanguageExt.Core/DataTypes/List/Lst.Internal.cs#L529-L559

This change occurred between versions 3.0.19 and 3.0.20.  Consistent with that is the fact that this test also fails for 3.0.20 but passes for 3.0.19.

**Comment by bender2k14** [comment]
FYI: The test I posted above passes on your new commit https://github.com/louthy/language-ext/commit/11c198c766719d5902b2e4faadda53af9b2baf26.

**Comment by louthy** [comment]
> The problem is that Lst<> is not immutable

It is immutable, there was a bug in the `InsertMany` method which builds a sub-node for the insertion separately before returning a transformed.  This was done for performance reasons, and I'll return to it later to make it work properly, but for now I've fixed it to be safe rather than fast.

@colethecoder  This is fixed in `v3.3.42`

**Comment by louthy** [comment]
Ok, yep, there was another issue which is now fixed and will be updated in `3.3.43`

**Comment by louthy** [comment]
`v3.3.43` on nu-get now
